src/algebra/direct_sum/decomposition.lean

```diff
@@ -72,6 +72,22 @@ def decompose : M ≃ ⨁ i, ℳ i :=
   left_inv := decomposition.left_inv,
   right_inv := decomposition.right_inv }
 
+protected lemma decomposition.induction_on {p : M → Prop}
+  (h_zero : p 0) (h_homogeneous : ∀ {i} (m : ℳ i), p (m : M))
+  (h_add : ∀ (m m' : M), p m → p m' → p (m + m')) : ∀ m, p m :=
+begin
+  let ℳ' : ι → add_submonoid M :=
+    λ i, (⟨ℳ i, λ _ _, add_mem_class.add_mem, zero_mem_class.zero_mem _⟩ : add_submonoid M),
```

Comment:
This is another argument for addressing #15053 sooner rather than later; but no reason to cancel merging this PR.